### ISSUE TITLE
Assignments in class body don't narrow type

### ISSUE THREAD

**Issue by ConsciousCode** [microsoft/pyright#6510]
# Assignments in class body don't narrow type
Note: if you are reporting a wrong signature of a function or a class in the standard library, then the typeshed tracker is better suited for this report: https://github.com/python/typeshed/issues.

If you have a question about typing or a behavior that you’re seeing in Pyright (as opposed to a bug report or enhancement request), consider posting to the [Pyright discussion forum](https://github.com/microsoft/pyright/discussions).

**Describe the bug**
Assignment of member attributes within the class body is ignored by type checking. It is not ignored when the assignment is outside of the class body when accessing the class (eg `Toy.method.dependent`) but not when accessing it on instances (eg `Toy().method.dependent`). I will be using the word "incorrect" for brevity, but it's more incomplete type inference; at no point is the type literally incorrect, but rather inferred as `Any` when a more specific type can be reasonably inferred.

**Code or Screenshots**
Incorrect behavior:
```python
class Toy:
    def method(self): ...
    def method_dependent(self): ...
    independent = method # Inferred correctly
    method.dependent = method_dependent # Type narrowing ignored

reveal_type(Toy.method) # (self: Toy) -> None
reveal_type(Toy.independent) # (self: Toy) -> None
#reveal_type(Toy.method.not_an_attr) # Any - correct, though maybe should be Unknown. AttributeError in Python
reveal_type(Toy.method.dependent) # Any - incorrect, should be narrowed to (self: Toy) -> None
```

Correct behavior:
```python
class Toy:
    def method(self): ...
    def method_dependent(self): ...
# Assignment after body
Toy.independent = Toy.method # Cannot assign member, OK
Toy.method.dependent = Toy.method_dependent # No error

reveal_type(Toy.method) # (self: Toy) -> None
reveal_type(Toy.independent) # (self: Toy) -> None
#reveal_type(Toy.method.not_an_attr) # Any, AttributeError in Python
reveal_type(Toy.method.dependent) # (self: Toy) -> None
```

Incorrect for both:
```python
reveal_type(Toy.method.dependent) # (self: Toy) -> None, correct
reveal_type(Toy().method.dependent) # Any, should be (self: Toy) -> None

# Sanity checks: Bound methods pass through function attributes
print(Toy().method) # <bound method Toy.method of <Toy object>>
print(Toy().method.dependent) # <function Toy.method_dependent>
```

So attributes added to *methods* in general don't narrow the type for instances. The following *does* work correctly:
```python
def method(): ...
def method_dependent(): ...
method.dependent = method_dependent
reveal_type(method.dependent) # () -> None
```

It's also worth noting that in VSCodium when I add `del method_dependent` (either to the end of the body or afterwards), it shows both `Toy.method_dependent` and `Toy().method_dependent` as autocomplete suggestions, and in no circumstances suggests 'Toy().method.dependent`. However, I'm not sure if this is Pyright or VSCodium.

There are workarounds I can use, like dedicated classes for these "methods", but this is valid Python so it ought to work. It also occurs to me that my original reason for doing this isn't quite valid, since `self.method.dependent()` wouldn't receive self and would require more machinery, but I figure this edge case is still technically incorrect despite not being particularly common and fixing it may make more common patterns better.

**VS Code extension or command-line**
Pyright 1.1.336 in VSCodium 1.84.2 x64 Ubuntu 23.04
```bash
consciouscode@theseus:~$ codium --version
1.84.2
1fc52b725b3f76b97eadfd988cea42a739ae923f
x64
```

**Comment by erictraut** [comment]
The current behavior is as designed, so I don't consider this a bug.

Pyright does not track attributes of function objects and generally considers it a type violation if you attempt to read or write an attribute on a function or method object. If you enable the `reportFunctionMemberAccess` rule, you'll see that it emits errors in this case. 

While the Python runtime does allow attributes to be set on function objects, it's an uncommon practice. The Python type system has little or no support for this.

If you want to do this, you could implement this in a type-safe manner using a wrapper object.

Code sample in [pyright playground](https://pyright-play.net/?code=MQAgDgngTglg5gCwC4C4RQKZgPZSQMQFcA7AYyRm2IFkMBbAIwygEFTSMBnTgXiSkIYAUEIBmUbHRBIIYGMTggYdHHhAtiEADQgAwgEMANof0NDwoaRPcQtJAmwATAOox7AESwZij70gDaAFSBAAo6AEoAuihCIHEgAPqiJKRoBsam5v5hIFGx8b5g3r7EqHpGJmYY-gB0dToaEJEi8SC%2BookJ8m4JCQAUnBiGojrJZGkVmdU5UQCUIAC0AHwgAHJUGDGtrYPDNUkpIDwgY6QtBRgdvaQVvQNDIyCB%2BlBwnGghNS9vOsEA1gB3b7vECfQHA%2BbLXJbbZxTBIQhQYggXaifanPrPV6cX6BcHY2YiKz6GwAFWwEBhcQAAnYHC43AhPEUfH58nF2iA6Bh7E57sNZlTtnUaucOZcuTz6QlCsU-PzRIL2cK6mKlKyWSUkEdJbzHMruXqarLWaUdYbpSatSJMAA3DBGBIyIp9ckQGoWpyEu0OwxO2QYV0UmryK1%2Bb0Ye2O52Bt0eqVOY1eU1IQlAA)

```python
# pyright: reportFunctionMemberAccess=true

from typing import Any, Callable

class MethodWithDependent[**P, R]:
    _func: Callable[P, R]
    dependent: Callable[..., Any]

    def __init__(self, func: Callable[P, R]) -> None:
        self._func = func

    def __call__(self, *args: P.args, **kwargs: P.kwargs) -> R:
        return self._func(*args, **kwargs)

class Toy:
    @MethodWithDependent
    def method(self):
        ...

    def method_dependent(self):
        ...

    independent = method
    method.dependent = method_dependent

reveal_type(Toy.method)
reveal_type(Toy.independent)
reveal_type(Toy.method.dependent)
```
